File: src/helpers/shallow-equal.ts
import Object from "@rbxts/object-utils";

/**
 * Compares two arbitrary values for shallow equality. Object values are compared based on their keys, i.e. they must
 * have the same keys and for each key the value must be equal.
 */
export function shallowEqual<T>(left: T, right: unknown): boolean {
	if (left === right) return true;

	if (!typeIs(left, "table") || !typeIs(right, "table")) return false;

	const keysLeft = Object.keys(left) as (keyof T)[];
	const keysRight = Object.keys(right) as (keyof typeof right)[];

	if (keysLeft.size() !== keysRight.size()) return false;

	return keysLeft.every((value, index) => value === right[index as never]);
}

File: src/utils/shallowEqual.lua
local function shallowEqual(left, right)
	if left == right then
		return true
	end

	if type(left) ~= "table" or type(right) ~= "table" then
		return false
	end

	if #left ~= #right then
		return false
	end

	for key, value in pairs(left) do
		if right[key] ~= value then
			return false
		end
	end

	return true
end

return shallowEqual
